### DISCUSSION TITLE
Line-height for Text element

### DISCUSSION THREAD

**Discussion by a-erlandsen-z** [slint-ui/slint#5920] [Ideas]
As far as I can see, there is no way to define line-height for a Text-element?

This would be very useful and help achieve better looking text display.

I'd expect more or less the same implementation as for the `line-height` CSS element.

See also: https://en.wikipedia.org/wiki/Leading

Bonus feature: allow vertical placement of a text element based on its baseline. The goal would be to align text elements in a more visually pleasing way.

**Comment by tronical** [comment]
Thanks for bringing this up here as discussion.

I'm hesitant about introducing `line-height` as I think that would come
along with the expectation of introducing the half-leading model. At the
moment we don't use that. The font's line-gap property is applied as a gap,
instead of the CSS way where half of it is distributed to the ascender
and the other half to the descender (half leading).

So, _if_ we were to introduce line-height, then it would have to come
along with a `half-leading` property:

```slint
Text {
    line-height: 24px;
    half-leading: true;
    font-size: 18px;
}
```

But that feels like a step backward and perhaps there's a better way to
address the actual issues of alignment.

I can see two different directions:

1. We introduce ways to query font metrics (such as the line gap,
ascender, descender, or capheight). With these and line gap only applied
between actual lines of text, maybe it's possible to compute the
necessary positions for the desired optical alignment you seek?

2. We introduce certain alignment/trim/edge properties to achieve the
designed optical effect without explicit computation.


(1) is relatively easy to implement, I think, while (2) requires a much
deeper understanding of the individual alignment issues.

Would you be able to share a few of the concrete text alignment issues you've
encountered?

**Reply by a-erlandsen-z** [reply]
Thanks for your reply.

I'm curious: what's the reasoning behind wanting to avoid the half-leading model? I will admit I'm coming at this from a designer's perspective and not a programmer's, but I look at this page: https://www.w3.org/TR/css-inline-3/ and I think: **Yes!** I want these features for laying out text.

I'm hesitant about your option 2, as that feels like we're trying to manually create metrics and positioning on top of metrics that are already available within the font files themselves - specifically put in place by the font designers, no?

I made a (very!) quick example here:
[SlintPad](https://slintpad.com/?snippet=export+component+Demo+inherits+Rectangle+%7B%0A++++width%3A+400px%3B%0A++++height%3A+400px%3B%0A%0A++++VerticalLayout+%7B%0A++++++++spacing%3A+10px%3B%0A%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+60px%3B%0A++++++++++++++++background%3A+%23cccccc%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++text%3A+%221%22%3B%0A++++++++++++++++++++font-size%3A+35px%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+45px%3B%0A++++++++++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++x%3A+10px%3B%0A++++++++++++++++++++width%3A+parent.width+-+self.x%3B%0A++++++++++++++++++++text%3A+%22This+is+the+first+entry%22%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++font-size%3A+22px%3B%0A++++++++++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+60px%3B%0A++++++++++++++++background%3A+%23cccccc%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++text%3A+%222%22%3B%0A++++++++++++++++++++font-size%3A+35px%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+45px%3B%0A++++++++++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++x%3A+10px%3B%0A++++++++++++++++++++width%3A+parent.width+-+self.x%3B%0A++++++++++++++++++++text%3A+%22The+second+element+is+too+long+for+a+single+line%22%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++font-size%3A+22px%3B%0A++++++++++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+60px%3B%0A++++++++++++++++background%3A+%23cccccc%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++text%3A+%223%22%3B%0A++++++++++++++++++++font-size%3A+35px%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+45px%3B%0A++++++++++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++x%3A+10px%3B%0A++++++++++++++++++++width%3A+parent.width+-+self.x%3B%0A++++++++++++++++++++text%3A+%22The+third+element%22%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++font-size%3A+22px%3B%0A++++++++++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+60px%3B%0A++++++++++++++++background%3A+%23cccccc%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++text%3A+%224%22%3B%0A++++++++++++++++++++font-size%3A+35px%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+45px%3B%0A++++++++++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++x%3A+10px%3B%0A++++++++++++++++++++width%3A+parent.width+-+self.x%3B%0A++++++++++++++++++++text%3A+%22The+fourth+element+is+so+long+it+needs+three+lines+to+fit+all+of+it%22%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++font-size%3A+22px%3B%0A++++++++++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+200px%3B%0A++++++++++++++++height%3A+100px%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++width%3A+parent.width%3B%0A++++++++++++++++++++height%3A+parent.height%3B%0A++++++++++++++++++++text%3A+%2212%22%3B%0A++++++++++++++++++++font-size%3A+36px%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.right%3B%0A++++++++++++++++++++vertical-alignment%3A+TextVerticalAlignment.bottom%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+100px%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++width%3A+parent.width%3B%0A++++++++++++++++++++height%3A+parent.height%3B%0A++++++++++++++++++++text%3A+%22Lorem+ipsum%22%3B%0A++++++++++++++++++++font-size%3A+21px%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++vertical-alignment%3A+TextVerticalAlignment.bottom%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D&style=native)

This is not beautiful typography, but is meant to show some of the issues I am facing and things that would be useful to laying out text more beautifully.

Here is a graphic representation of what I would have wanted the above SlintPad example to actually look like:

![Artboard 3](https://github.com/user-attachments/assets/a06e5e19-4c0d-4cbe-901b-0bb95b166516)

The first and third entries should be aligned optically to along the center line. Note that the optical weight of a line of text are not simply the precise center between top and bottom extremities.

The second entry shows how text is ending up outside the box when wrapping over a second line. Imagine a menu or list view where you want to adjust the leading / line height to fit inside a more limited space like here. Then when vertically centering I again want the optical center to come into effect so for example the ability to center based on the part between the baseline and x-height: https://www.w3.org/TR/css-inline-3/#text-edges

The fourth item shows another but related issue: Today, if you enable word-wrap on a text element, and you also enable overflow: elide, you get one line only. Often I want to show one or more lines of wrapped text, and what is left after that should be truncated (or clipped).

And below: a bonus item. Aligning by baseline. See again here for reference: https://www.w3.org/TR/css-inline-3/

**Reply by tronical** [reply]
Great examples. thanks!

Regarding the half-leading: The reason I'm hesitant about introducing it is because it would be a breaking change for single-line text and it just doesn't feel like a natural choice for layout. I mean, the font provides ascent, descent, and line gap metrics. The line gap is the leading and on the Wikipedia page you referred to it's well described. The CSS choice of cutting this "strip of lead" into half seems like another complication that requires explaining and mental gymnastics when laying out text. I'm curious what would be a motivation for doing so?

That said, we need to have better features in Slint for layout out text and gradually move towards a feature set that CSS has.

I'm a little glad that you're hesitant about option (2), so I think I'd like to proceed with option (1) at first. They're not mutually exclusive, but I think we can learn from the uses of (1) in the future and extract useful convenience.

Here's an example of how perhaps (1) could solve the optical alignment of the first and third entry:

(dramatically simplified code)
```slint
HorizontalLayout {
        // NEW: imagine a `Text` element had a `font-metrics` property.
        property <length> common-baseline: max(label1.font-metrics.ascent, label2.font-metrics.ascent);

        Rectangle {
                width: 60px;
                background: #cccccc;
                label1  := Text {
                    text "1";
                    font-size: 35px;
                    // NEW: imagine a `Text` element had a way of specifying the distance from the top of the `Text` box to the start of the baseline
                    baseline: root.common-baseline;
                }
            }
            Rectangle {
                height: 45px;
                background: #aaaaaa;

                label2 := Text {
                    x: 10px;
                    width: parent.width - self.x;
                    text: "This is the first entry";
                    font-size: 22px;
                    baseline: root.common-baseline;
                }
            }
        }
}
```

I have the guts feeling that could work and be done with little effort. _If_ it works, then such a baseline computation and application could be done automatically in the future, so that ideally we could have something like this:

(even more simplified again)
```slint
HorizontalLayout {
    /// implicitly generated by the compiler, invisible: property <length> common-baseline: computed-max-ascent-of-all-text-elements;
    Text {
        font-size: 22px;
        vertical-alignment: baseline;
        /// implicitly generated by the compiler: baseline: root.common-baseline;
    }
    Text {
        font-size: 45px;
        vertical-alignment: baseline;
        /// implicitly generated baseline binding here as well
    }
}
```

Does this sounds like a reasonable direction? I Think we could provide a `font-metrics` property that would pave the way to manually computed alignments, along with say a `baseline` property, relatively easily.

**Reply by a-erlandsen-z** [reply]
I'm liking your thoughts on a baseline alignment. If you want to expand further in the direction of optical centering, x-height would perhaps also be useful?

Would you then approach my initial line-height / leading question by just adjusting the distance between each baseline in a horizontal layout and/or multi-line block of text?

EDIT: I'm glad you agree better text layout features is a goal for slint. 
EDIT 2: In general, I'd hope to base metrics as much as possible on the intention of the font designer - so, essentially metrics that are available from the font file I guess?

**Reply by tronical** [reply]
Yes, `x-height` and `cap-height` would also make sense.

I'd indeed approach your initial question by adjusting the baseline in a horizontal layout. In multi-line text that would apply to the first line, and the subsequent baseline should perhaps be defined based on the descent, ascent, and line gap?

Yes, with metrics I'm really referring to the metrics the font designer placed into the font, without any other interpretation (such as dividing the line gap by half to turn it into leading :-)

**Reply by tronical** [reply]
I've created https://github.com/slint-ui/slint/issues/6047 and https://github.com/slint-ui/slint/issues/6048 as first conclusions from this (continuing) discussion.

**Reply by a-erlandsen-z** [reply]
Perfect!

**Reply by a-erlandsen-z** [reply]
Any thoughts on combining word-wrap and elide?

**Reply by ogoffart** [reply]
> Any thoughts on combining word-wrap and elide?

This should be already supported. You have to specify a height for the text: [example](https://slintpad.com/?snippet=export+component+Demo+inherits+Rectangle+%7B%0A++++width%3A+300px%3B%0A++++height%3A+200px%3B%0A%0A++++Rectangle+%7B%0A++++++++height%3A+60px%3B%0A++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++Text+%7B%0A++++++++++++width%3A+parent.width%3B%0A++++++++++++height%3A+parent.height%3B%0A++++++++++++text%3A+%22The+fourth+element+is+so+long+it+needs+three+lines+to+fit+all+of+it%22%3B%0A++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++vertical-alignment%3A+center%3B%0A++++++++++++font-size%3A+22px%3B%0A++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++overflow%3A+elide%3B%0A++++++++%7D%0A++++%7D%0A%0A%7D)

**Reply by a-erlandsen-z** [reply]
Ah. I'm sorry. I see that now.

I simply looked at the large amount of available space under item four in my example [(SlintPad)](https://slintpad.com/?snippet=export+component+Demo+inherits+Rectangle+%7B%0A++++width%3A+400px%3B%0A++++height%3A+400px%3B%0A%0A++++VerticalLayout+%7B%0A++++++++spacing%3A+10px%3B%0A%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+60px%3B%0A++++++++++++++++background%3A+%23cccccc%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++text%3A+%221%22%3B%0A++++++++++++++++++++font-size%3A+35px%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+45px%3B%0A++++++++++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++x%3A+10px%3B%0A++++++++++++++++++++width%3A+parent.width+-+self.x%3B%0A++++++++++++++++++++text%3A+%22This+is+the+first+entry%22%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++font-size%3A+22px%3B%0A++++++++++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+60px%3B%0A++++++++++++++++background%3A+%23cccccc%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++text%3A+%222%22%3B%0A++++++++++++++++++++font-size%3A+35px%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+45px%3B%0A++++++++++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++x%3A+10px%3B%0A++++++++++++++++++++width%3A+parent.width+-+self.x%3B%0A++++++++++++++++++++text%3A+%22The+second+element+is+too+long+for+a+single+line%22%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++font-size%3A+22px%3B%0A++++++++++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+60px%3B%0A++++++++++++++++background%3A+%23cccccc%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++text%3A+%223%22%3B%0A++++++++++++++++++++font-size%3A+35px%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+45px%3B%0A++++++++++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++x%3A+10px%3B%0A++++++++++++++++++++width%3A+parent.width+-+self.x%3B%0A++++++++++++++++++++text%3A+%22The+third+element%22%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++font-size%3A+22px%3B%0A++++++++++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+60px%3B%0A++++++++++++++++background%3A+%23cccccc%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++text%3A+%224%22%3B%0A++++++++++++++++++++font-size%3A+35px%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+45px%3B%0A++++++++++++++++background%3A+%23aaaaaa%3B%0A%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++x%3A+10px%3B%0A++++++++++++++++++++width%3A+parent.width+-+self.x%3B%0A++++++++++++++++++++height%3A+parent.height%3B%0A++++++++++++++++++++text%3A+%22The+fourth+element+is+so+long+it+needs+three+lines+to+fit+all+of+it%22%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++font-size%3A+22px%3B%0A++++++++++++++++++++wrap%3A+word-wrap%3B%0A++++++++++++++++++++overflow%3A+elide%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++width%3A+200px%3B%0A++++++++++++++++height%3A+100px%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++width%3A+parent.width%3B%0A++++++++++++++++++++height%3A+parent.height%3B%0A++++++++++++++++++++text%3A+%2212%22%3B%0A++++++++++++++++++++font-size%3A+36px%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.right%3B%0A++++++++++++++++++++vertical-alignment%3A+TextVerticalAlignment.bottom%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+100px%3B%0A++++++++++++++++Text+%7B%0A++++++++++++++++++++width%3A+parent.width%3B%0A++++++++++++++++++++height%3A+parent.height%3B%0A++++++++++++++++++++text%3A+%22Lorem+ipsum%22%3B%0A++++++++++++++++++++font-size%3A+21px%3B%0A++++++++++++++++++++horizontal-alignment%3A+TextHorizontalAlignment.left%3B%0A++++++++++++++++++++vertical-alignment%3A+TextVerticalAlignment.bottom%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D&style=native) and just assumed this had still not been implemented. Good to see.

So this simply points back to the original topic of this conversation, then: adjusting line space will solve this as well.

Thank you.
